protocol.html

```diff
@@ -559,6 +572,44 @@ <h3 property="schema:name">Resource Containment</h3>
+
+                  <section id="contained-resource-metadata" inlist="" rel="schema:hasPart" resource="#contained-resource-metadata">
+                    <h4 property="schema:name">Contained Resource Metadata</h4>
+                    <div datatype="rdf:HTML" property="schema:description">
+                      <p>Container descriptions are not limited to containment triples. To further support client navigation and application interaction, servers can include <a href="#resource-metadata">resource metadata</a> about contained resources as part of the container description, as described below.</p>
+
+                      <p><span about="" id="server-contained-resource-metadata" rel="spec:requirement" resource="#server-contained-resource-metadata"><span property="spec:statement"><span rel="spec:requirementSubject" resource="spec:Server">Servers</span> <span rel="spec:requirementLevel" resource="spec:MUST">MUST</span> include <a href="#contained-resource-metadata-statements" rel="rdfs:seeAlso">resource metadata about contained resources</a> as part of the container description, unless that information is inapplicable to the server.</span></span></p>
+
+                      <p id="contained-resource-metadata-statements" about="#contained-resource-metadata-statements" typeof="skos:Collection"><span property="skos:prefLabel">Contained resource metadata statements</span> include the properties:</p>
+
+                      <dl about="#contained-resource-metadata-statements" rel="skos:member">
+                        <dt about="#contained-resource-metadata-rdf-type" id="contained-resource-metadata-rdf-type" property="skos:prefLabel"><code>rdf:type</code></dt>
+                        <dd about="#contained-resource-metadata-rdf-type" property="skos:definition">A class whose URI is the expansion of the <em>URI Template</em> [<cite><a class="bibref" href="#bib-rfc6570">RFC6570</a></cite>] <code>http://www.w3.org/ns/iana/media-types/{+iana-media-type}#Resource</code>, where <code>iana-media-type</code> corresponds to a value from the IANA Media Types [<cite><a class="bibref" href="#bib-iana-media-types">IANA-MEDIA-TYPES</a></cite>].</dd>
+                        <dt about="#contained-resource-metadata-stat-size" id="contained-resource-metadata-stat-size" property="skos:prefLabel"><code>stat:size</code></dt>
+                        <dd about="#contained-resource-metadata-stat-size" property="skos:definition">A non-negative integer giving the size of the resource in bytes.</dd>
+                        <dt about="#contained-resource-metadata-dcterms-modified" id="contained-resource-metadata-dcterms-modified" property="skos:prefLabel"><code>dcterms:modified</code></dt>
+                        <dd about="#contained-resource-metadata-dcterms-modified" property="skos:definition">The date and time when the resource was last modified.</dd>
```

Comment:
Assuming that a server provides a `Last-Modified` header and assuming that we do not want to break how caching works for browsers, I do not see how this (`dcterms:modified` or `stat:mtime`) can be implemented without causing a cascade of updates to the root of a Storage.

For example, consider a resource at `/lvl-1/lvl-2/foo.ttl` with the following response headers:

```http
Last-Modified: Sun, 05 Dec 2021 17:45:02 GMT
ETag:  "abcd"
```

The container at `/lvl-1/lvl-2/` has the following headers:

```http
Last-Modified: Tue, 07 Dec 2021 12:23:15 GMT
ETag: "1234"
```

That container would include the following triples:

```turtle
</lvl-1/lvl-2/> 
    dcterms:modified "2021-12-07T12:23:15Z"^^xsd:dateTime ;
    ldp:contains </lvl-1/lvl-2/foo.ttl> .
</lvl-1/lvl-2/foo.ttl>
    dcterms:modified "2021-12-05T17:45:02Z"^^xsd:dateTime .
```

The container at `/lvl-1/` has the following headers:

```http
Last-Modified: Tue, 07 Dec 2021 19:37:22 GMT
ETag: "zyxw"
```

And the following body:

```turtle
</lvl-1/> 
    dcterms:modified "2021-12-07T19:37:22Z"^^xsd:dateTime ;
    ldp:contains </lvl-1/lvl-2/> .
</lvl-1/lvl-2/>
    dcterms:modified "2021-12-07T17:45:02Z"^^xsd:dateTime .
```

Now consider that a client adds a triple to `/lvl-1/lvl-2/foo.ttl`. The result will lead to the `Last-Modified` being updated on `/lvl-1/lvl-2/foo.ttl`. The content of that resource has changed, so the ETag header also changes.

Because the `Last-Modified` header changes, the representation of `/lvl-1/lvl-2/` now also changes to reflect the new status of the contained resource. As a consequence, and in order to ensure that clients receive the latest version of `/lvl-1/lvl-2/`, the `Last-Modified` and `ETag` headers of this container resource must change.

Because the `Last-Modified` header for `/lvl-1/lvl-2/` changes, the representation of `/lvl-1/` also needs to change, which leads to a further cascade of changes to the root of the storage.

There are three ways around this, as I see it:

* Do not include the `dcterms:modified` or `stat:mtime` triples (i.e. ignore this part of the spec)
* Don't worry about breaking how browser caching works (i.e. ignoring app/client needs)
* Put this data in a separate (auxiliary) resource that doesn't lead to a cascade of change.

I prefer the third option because it makes it possible to include this data without breaking how browsers interact with HTTP resources, but absent that approach, the only route forward I see (while staying in line with this spec) is to simply _not_ include the modified time of contained resources.

Reply:
Yes, @acoburn , I have been arguing the same, an auxiliary resource solves all problems, it can represent the entirety of the resource (and it makes sense to add data for each representation too if needed), avoids cascading problems, can be subject to separate authorization, can be used to track any changes and should be at least as easy to implement. 

However, we haven't been able to find consensus around that for 0.9, so indeed, the result is that these data may be stale when conditional requests are used. I disagree with the conclusion, but as the RFC7232 is fairly vague at this point, I have chosen to accept it for 0.9, sincerely hoping we can revisit for 1.0.

Reply:
>Because the Last-Modified header changes, the representation of `/lvl-1/lvl-2/` now also changes to reflect the new status of the contained resource.

While the representation data of `/lvl-1/lvl-2/` changes because of the `dcterms:modified` value of `/lvl-1/lvl-2/foo.ttl` , the `Last-Modified` header of `/lvl-1/lvl-2/` need not change - nor prohibited - as there were no changes to the containment triples ( #server-container-last-modified ).

A recommendation (or advisement) along the following lines may be necessary:

When the resource metadata of an existing contained resource changes, the server MUST send a weak `entity-tag` when responding to container’s request URI.

---

I want to mention again that the requirements introduced by this PR does not expect a container (`/lvl-1/lvl-2/`) to include resource metadata about itself (`</lvl-1/lvl-2/> dcterms:modified "2021-12-07T12:23:15Z"^^xsd:dateTime .`). (I don't care if that's not the point - it easily introduces complications/assumptions to the discussion that we are better off without.)   And again, if resource metadata about the container itself is desired - I may have missed the discussion that calls for it - the specification should needs to say so because both the generation and protection of those resource metadata needs to apply. On a related note:

When a server plans to include resource metadata about the container in the response of the same container, the server should determine the `Last-Modified` header value as a regular change to the representation data. (It can send a strong `entity-tag` as usual.)

Reply:
This does mean that apps/users will see stale data. From an impl perspective (given the choice), I would rather not provide the data than provide it in a way that confuses users.

I am really trying to be constructive here, but I don't see how this is implementable, given this text

Reply:
As it stands, the Protocol uses the same requirement levels as in the RFC for Last-Modified and ETag headers. Servers might not include either Last-Modified or ETag headers, and so clients can't absolutely rely on them for being there.

That aside,

https://datatracker.ietf.org/doc/html/rfc7232#section-2.4 and https://datatracker.ietf.org/doc/html/rfc7232#section-6 suggests that headers with entity-tags will have higher precedence.

Server needs to enable the use of conditional requests so that client can eventually get a 200/304/412 or whatever.

Perhaps I'm missing something here but as mentioned in previous comment, I don't see a staleness issue around entity-tags.

If a client is only making decisions with Last-Modified/If-*-Since (when Last-Modified is provided by the server), then the representation metadata in the response will not help them to differentiate between a change to containment triples or any change to the representation. In this particular scenario, client has to make a GET request on the container.

We can certainly revisit that in 1.0.

For now, I suggest to add the following and go ahead with 0.9:

>When the resource metadata of an existing contained resource changes, the server MUST send a weak entity-tag when responding to container’s request URI.

Reply:
> When the resource metadata of an existing contained resource changes, the server MUST send a weak entity-tag when responding to container’s request URI.

I would like to just register that sending a weak ETag in this case does not address my concerns about browser caching at all. Any client sending a GET with an If-None-Match header will still receive a 304 (regardless of the weakness indicator) and therefore will present stale data to users

Reply:
The suggestion is derived directly from the RFCs and is not required to be said but figured it can help.

There will be a different entity-tag for all changes to the container, so I don't quite see how a client using entity-tag x meanwhile the current representation's entity-tag is y will return a 304.

Reply:
If the entity tag changes from x to y, there is no issue, but that is orthogonal to the weakness indicator

Reply:
Good. Then in what case would it not change from x to y?

Reply:
Any time the representation changes, the ETag changes. There is no argument about that. I am only responding to the proposed requirement to use weak ETags. And to be clear, I am not opposed to using weak ETags, but we can't claim  that weak ETags will solve this issue

Reply:
The weak entity-tag is suggested because strong entity-tag wouldn't be correct when only the resource metadata changes. All meanwhile allowing the client to make conditional requests. Right at this second, I don't have a strong opinion on adding that requirement because I do think it comes directly from the RFC but it may be reasonable to mention.

That aside, I acknowledge that there is (always) room to discuss/work this out further but it doesn't need to be a blocker for 0.9. (We mark handful of stuff to be revisited for 1.0, and I don't see why it can't be done here as well.)

Reply:
https://github.com/solid/specification/pull/352/commits/746520d614e1631334879e5f55b010665d707a1f